### PR TITLE
[mlir][DeadCodeAnalysis] Don't Require `RegionBranchTerminatorOpInterface` in `visitRegionTerminator()`

### PR DESCRIPTION
**Pull request by justinfargnoli**
Fix for a crash reported in #64975. 

The crash occurs in the cast located [here](https://github.com/llvm/llvm-project/blob/ece5dd101c7e4dc2fd23428abd312f75fd3d3eaf/mlir/lib/Analysis/DataFlow/DeadCodeAnalysis.cpp#L262) because `llvm.unreachable` doesn't implement `RegionBranchTerminatorOpInterface`. 

The crash is caused by `DeadCodeAnalysis` assuming that `isa<RegionBranchOpInterface>(op->getParentOp())` implies `isa<RegionBranchTerminatorOpInterface>(op)` in `DeadCodeAnalysis::visit()`. 

This patch tried to fix this by enabling the analysis to proceed regardless of whether `op` is a `RegionBranchTerminatorOpInterface`.

### CODE REVIEW

**Comment by justinfargnoli** [comment]
Note: I'm very new to MLIR. So, I'm not 100% confident this is the right solution. Please feel free to suggest entirely different ways of tackling this problem if my chosen route doesn't make sense.

File: mlir/test/Analysis/DataFlow/test-dead-code-analysis-regression.mlir
Comment (Mogball) [inline mlir/test/Analysis/DataFlow/test-dead-code-analysis-regression.mlir:3]: suggestion: please merge this test case with one of the existing ones for dataflow analysis.

Also, this seems like invalid IR. `scf.execute_region` should be verifying its terminators

File: mlir/lib/Analysis/DataFlow/DeadCodeAnalysis.cpp
Comment (Mogball) [inline mlir/lib/Analysis/DataFlow/DeadCodeAnalysis.cpp:390]: nice

**Review by Mogball — APPROVED** [review]
Thanks for the fix! LGTM after merging the test cases

**Pull request merged**
